### PR TITLE
chore: implement semantics cluster_by

### PR DESCRIPTION
**Pull request by chelsea-lin**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: tests/system/large/operations/test_semantics.py [hunk 1]
```diff
@@ -130,6 +130,68 @@ def test_agg_invalid_cluster_column_raise_error(gemini_flash_model, cluster_colu
...
 
+@pytest.mark.parametrize(
+    ("n_clusters"),
+    [
+        pytest.param(1, id="one", marks=pytest.mark.xfail(raises=ValueError)),
+        pytest.param(2, id="two"),
+        pytest.param(4, id="four"),
+    ],
+)
+def test_cluster_by_w_n_clusters(session, text_embedding_generator, n_clusters):
```
Comment (sycai) [inline tests/system/large/operations/test_semantics.py:141]: nit: just "test_cluster_by" should suffice. The varying numbers of clusters are already reflected in the function parameter

**Review by sycai — APPROVED** [review]

File: tests/system/large/operations/test_semantics.py (hunk 1 above)
Comment (chelsea-lin) [reply tests/system/large/operations/test_semantics.py:141]: Good to know. Thanks!

File: bigframes/operations/semantics.py [hunk 2]
```diff
@@ -194,6 +194,88 @@ def agg(
...
+
+            >>> df = bpd.DataFrame({
+            ...     "Product": ["Smartphone", "Laptop", "T-shirt", "Jeans"],
+            ... })
+            >>> df.semantics.cluster_by("Product", "Cluster ID", model, n_clusters=2)
+                    Product  Cluster ID
+            0    Smartphone           2
+            1        Laptop           2
+            2       T-shirt           1
+            3         Jeans           1
```
Comment (TrevorBergeron) [inline bigframes/operations/semantics.py:229]: is this deterministic - or are cluster centroids initialized randomly?

Comment (chelsea-lin) [reply bigframes/operations/semantics.py:229]: Though initial tests suggest consistent results, I'm not fully convinced. I'll continue monitoring the test outcomes. If inconsistencies arise, I'll adjust the verification process to focus solely on the number of unique IDs.

**Pull request merged**